sklearn/metrics/regression.py

```diff
@@ -314,16 +431,40 @@ def r2_score(y_true, y_pred, sample_weight=None):
     else:
         weight = 1.
 
-    numerator = (weight * (y_true - y_pred) ** 2).sum(dtype=np.float64)
+    numerator = (weight * (y_true - y_pred) ** 2).sum(axis=0,
+                                                      dtype=np.float64)
     denominator = (weight * (y_true - np.average(
-        y_true, axis=0, weights=sample_weight)) ** 2).sum(dtype=np.float64)
-
-    if denominator == 0.0:
-        if numerator == 0.0:
-            return 1.0
-        else:
-            # arbitrary set to zero to avoid -inf scores, having a constant
-            # y_true is not interesting for scoring a regression anyway
-            return 0.0
-
-    return 1 - numerator / denominator
+        y_true, axis=0, weights=sample_weight)) ** 2).sum(axis=0,
+                                                          dtype=np.float64)
+    nonzero_denominator = denominator != 0
+    nonzero_numerator = numerator != 0
+    valid_score = nonzero_denominator & nonzero_numerator
+    output_scores = np.ones([y_true.shape[1]])
+    output_scores[valid_score] = 1 - (numerator[valid_score] /
+                                      denominator[valid_score])
+    # arbitrary set to zero to avoid -inf scores, having a constant
+    # y_true is not interesting for scoring a regression anyway
+    output_scores[nonzero_numerator & ~nonzero_denominator] = 0.
+    if multioutput is None:
+        # @FIXME change in 0.18
+        warnings.warn("Default 'multioutput' behavior now corresponds to "
+                      "'variance_weighted' value, it will be changed "
+                      "to 'uniform_average' in 0.18.",
+                      DeprecationWarning)
+        multioutput = 'variance_weighted'
+    if multioutput == 'raw_scores':
+        # return scores individually
+        return output_scores
+    elif multioutput == 'uniform_average':
+        # passing None as weights results is uniform mean
+        multioutput = None
+    elif multioutput == 'variance_weighted':
+        multioutput = denominator
+        # avoid fail on constant y or one-element arrays
+        if not any(nonzero_denominator):
```

Comment:
np.any is OK.

np.all(nonzero_numerate) has different semantics:
not (x OR y) ~ (not x) AND (not y) != x AND y